### DISCUSSION TITLE
Does the googlecloud / stackdriver support metrics and traces from opentelemetry?

### DISCUSSION THREAD

**Discussion by tyrone-anz** [open-telemetry/opentelemetry-collector-contrib#2948] [Q&A]
We've been exporting our opentelemetry metrics to the collector (v0.23.0) using `opentelemetry-go` (v0.17.0) and the collector is configured to forward to google cloud. However, we are not successful in doing so and the collector is getting errors.

Depending on the metric distribution, it complains about a different error. This is a sample error from the collector.

```
2021-03-31T01:05:36.094Z	info	exporterhelper/queued_retry.go:276	Exporting failed. Will retry the request after interval.	{"component_kind": "exporter", "component_type": "googlecloud", "component_name": "googlecloud", "error": "rpc error: code = InvalidArgument desc = Field timeSeries[0].points[0].distributionValue had an invalid value: Distribution value has |bucket_counts| fields that sum to 0 which does not equal the |count| field value of 847.", "interval": "4.98550433s"}
```

`Distribution value has |bucket_counts| fields that sum to 0 which does not equal the |count| field value of 847`

When I inspected the googlecloud exporter code, I can see it is using opencensus libraries to send the metrics. Based on that, I am beginning to think that it can only export metrics if it from the opencensus receiver. I cannot find any documentation about it though. 

My assumption before is any exporter in the collector can forward successfully any telemetry data the collector is receiving regardless of the receiver used. But in my experience with the googlecloud exporter, we are only able to forward metrics successfully if from opencensus receiver and not otlp receiver. Or is this just a matter of the otlp receiver not yet stable?

**Comment by dashpole** [comment]
Apologies for the slow response.
The googlecloud exporter export metrics from any receiver.  Since the OTel metrics API/SDK isn't stable yet, our opentelemetry metrics exporter can't be stable, so the collector still uses the opencensus exporter.  It should still be able to send metrics from opentelemetry.

This is likely a bug in the google cloud opencensus exporter, or the go SDK (but probably in our exporter).

If you can help us reproduce the issue, that would be helpful.  I can think of a few approaches:

1. It would be helpful to know the details of metric that is causing problems.  If you can enable the `logging` exporter (with `logLevel: debug` along-side the google-cloud one, we can figure out what the bucket counts are, and such. With that, we might be able to construct an easy repro.
2. If you already have a reproduction using opentelemetry-go that you can share, that would also work.

**Reply by tyrone-anz** [reply]
Thanks for the response, @dashpole. I will prepare some basic setups that replicate this. I have not looked at this issue for a while but I think it is still a problem at v0.19.0 - the latest one we use. I will share when ready.

**Reply by tyrone-anz** [reply]
Hi @dashpole, I revisited this issue today.

I upgraded our `opentelemetry-go` SDK to `v1.0.0-RC1` (`v0.21.0` for metrics) and run the collector on `v0.31.0`. The issue with `bucket counts` has been fixed. From investigation, it seems due to the old collector exporter which has now been replaced on `v0.21.0`.

However, the following issue is now happening after the bucket issue is resolved.

```
One or more TimeSeries could not be written: Field timeSeries[1] had an invalid value: Duplicate TimeSeries encountered
```

This issue still causes our metrics not to get uploaded to google cloud. Some metrics may still get uploaded but once those errors accumulate on the collector, no good metric gets uploaded anymore.
 
My theory is the googlecloud exporter is maybe ignoring the individual attribute of each data as such the data points look the same. 

On the sample code below, I recorded two data with varying `rpc.method` values. You can see on the collector proto request log (available at the bottom of the file) that it was sent correctly with varying attributes but the collector log does not show this and eventually throws the `Duplicate Timeseries` error.

https://github.com/tyrone-anz/export-otlp-googlecloud/blob/main/main.go

I already tried all different selector aggregators (exact, inexpensive, histogram) to get around the issue but none helps. 

Please have a look when able and let me know if there more info to help in the investigation. Thanks!

**Reply by dashpole** [reply]
I am able to reproduce the issue with collector version v0.31.0 (although I had to fill in some values for the `batch` and `memory_limiter` processors in your config).  With the latest version (v0.34.0), I don't see the error anymore, but the "Data point attributes" are also removed from the printed-out metric.

I removed the batch processor, and was still able to reproduce the issue, so that isn't it.

**Reply by dashpole** [reply]
Can you try disabling retry on failure for the exporter?
```yaml
exporters:
  googlecloud:
    retry_on_failure:
      enabled:false
```

Right now, the `googlecloud` exporter doesn't handle partial failures well.  If any of the metrics fail to write, the monitoring API returns an error code indicating the failure.  Retrying a partially failed request will always fail.

The odd thing is: When I disable retry_on_failure, everything seems to work normally.  I would have expected that to uncover the "root" error, but it seems to just work.  I checked the metric explorer in my project, and the metric is appearing as I would expect.

**Reply by tyrone-anz** [reply]
Thanks for the reply @dashpole. Will let you know how it goes. If you did not lose any metric data, then that's good.

**Reply by tyrone-anz** [reply]
Hi @dashpole, I did some testing today with `v0.34.0` with retry disabled. It works but only if aggregator is set at `histogram distribution`. If I switched to `exact distribution`, the duplicate timeseries error is still there. 

I think there is issue with exporting value recorder metrics to googlecloud. On the googlecloud exporter (not through collector) link below, it has a custom aggregator `NewWithCloudMonitoringDistribution()` that just get the last value. It is a good workaround but many captured data will be lost and I am already exporting these metrics to somewhere else without issue.

https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/9616b5040d4ee15f0f7e623211c2f145684dfefb/exporter/metric/selector.go#L47

Do you think there is an issue with the otlp googlecloud exporter if the aggregator is exact distribution?

Here is a log when using the `exact distribution`

```
2021-09-20T14:14:55.641+1000    INFO    loggingexporter/logging_exporter.go:56  MetricsExporter {"#metrics": 1}
2021-09-20T14:14:55.642+1000    DEBUG   loggingexporter/logging_exporter.go:66  ResourceMetrics #0
Resource labels:
     -> service.name: STRING(unknown_service:___go_build_main_go)
     -> telemetry.sdk.language: STRING(go)
     -> telemetry.sdk.name: STRING(opentelemetry)
     -> telemetry.sdk.version: STRING(1.0.0-RC1)
InstrumentationLibraryMetrics #0
InstrumentationLibrary  
Metric #0
Descriptor:
     -> Name: test.dummy.one
     -> Description: 
     -> Unit: 
     -> DataType: Gauge
NumberDataPoints #0
Data point attributes:
     -> rpc.method: STRING(Hi)
StartTimestamp: 2021-09-20 04:14:53.48167 +0000 UTC
Timestamp: 2021-09-20 04:14:55.48552 +0000 UTC
Value: 10
NumberDataPoints #1
Data point attributes:
     -> rpc.method: STRING(Hi2)
StartTimestamp: 2021-09-20 04:14:53.48167 +0000 UTC
Timestamp: 2021-09-20 04:14:55.48552 +0000 UTC
Value: 20
NumberDataPoints #2
Data point attributes:
     -> rpc.method: STRING(Hi3)
StartTimestamp: 2021-09-20 04:14:53.48167 +0000 UTC
Timestamp: 2021-09-20 04:14:55.48552 +0000 UTC
Value: 30
NumberDataPoints #3
Data point attributes:
     -> rpc.method: STRING(Hi4)
StartTimestamp: 2021-09-20 04:14:53.48167 +0000 UTC
Timestamp: 2021-09-20 04:14:55.48552 +0000 UTC
Value: 40
NumberDataPoints #4
Data point attributes:
     -> rpc.method: STRING(Hi5)
StartTimestamp: 2021-09-20 04:14:53.48167 +0000 UTC
Timestamp: 2021-09-20 04:14:55.48552 +0000 UTC
Value: 50

2021-09-20T14:15:00.572+1000    error   exporterhelper/queued_retry.go:254      Exporting failed. Try enabling retry_on_failure config option.  {"kind": "exporter", "name": "googlecloud", "error": "rpc error: code = Internal desc = One or more TimeSeries could not be written: Internal error encountered. Please retry after a few seconds. If internal errors persist, contact support at https://cloud.google.com/support/docs.: timeSeries[0-4]"}
go.opentelemetry.io/collector/exporter/exporterhelper.(*retrySender).send
        go.opentelemetry.io/collector@v0.34.0/exporter/exporterhelper/queued_retry.go:254
go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsSenderWithObservability).send
        go.opentelemetry.io/collector@v0.34.0/exporter/exporterhelper/metrics.go:114
go.opentelemetry.io/collector/exporter/exporterhelper.(*queuedRetrySender).start.func1
        go.opentelemetry.io/collector@v0.34.0/exporter/exporterhelper/queued_retry.go:155
go.opentelemetry.io/collector/exporter/exporterhelper/internal.ConsumerFunc.Consume
        go.opentelemetry.io/collector@v0.34.0/exporter/exporterhelper/internal/bounded_queue.go:103
go.opentelemetry.io/collector/exporter/exporterhelper/internal.(*BoundedQueue).StartConsumersWithFactory.func1
        go.opentelemetry.io/collector@v0.34.0/exporter/exporterhelper/internal/bounded_queue.go:82
```

**Reply by dashpole** [reply]
I'm continuing to look into this, but in the meantime @ymotongpoo, who added NewWithCloudMonitoringDistribution() may be able to answer questions about the distribution.

**Reply by dashpole** [reply]
There seems to be a limitation around what can be in a single request sent to cloud monitoring.  Even if metrics have different timestamps, it seems to be rejected if they come in a single request.  If I change the example from using label values: [Hello, Hi, Hi, Hi, Hi] to [Hello, Hi, Foo, Bar, Baz], and remove the batch processor, the error goes away and I can see metrics.  But if I add the batch processor to the collector, I get the error again.  You can see my modifications to your repro here: https://github.com/dashpole/export-otlp-googlecloud/tree/repro.

Looking at some other internal bugs, this seems to be an issue similar to others we've seem before.  I'll follow up with them and circle back once I have a resolution.

**Reply by dashpole** [reply]
I've learned a few things that should help you resolve this:

First, Google Cloud Monitoring doesn't let you send the same metric (with same labels) twice in a single batch of metrics.  We will need to work around this limitation, at least for now.

Second, `NewWithExactDistribution()` converts the metric from a histogram to a gauge before sending to cloud monitoring.  So when you do:
```golang 
        cont := controller.New(processor.New(selector.NewWithExactDistribution(), exporter),
        ...
	valuerecorder.Record(ctx, 20, attribute.Any("rpc.method", "Hi"))
	valuerecorder.Record(ctx, 20, attribute.Any("rpc.method", "Hi"))
```
you are sending two gauges with the same labels.  If those get sent in the same batch to cloud monitoring, it will be rejected as you showed above.  For histograms, you should use `NewWithHistogramDistribution()`, which will keep it as a distribution when sending to GCM.

Third, using the `batch` processor it is still possible to combine two histogram metrics in the same batch depending on your client's `CollectPeriod` and your batch `timeout`, which you don't want to do.  For example, if your batch timeout is 60 seconds, and your CollectPeriod is 2 seconds, you could (depending on how often the metric itself is recorded) send as many as 30 histograms in the same batch of metrics.  To be safe, you should make sure the batch timeout is less than or equal to the batch timeout.  If you do this you _can_ still record metrics frequently (e.g. every second), so long as CollectPeriod isn't too short.  For example, if you `Record` a metric 10 times in a single CollectPeriod, it will only send a single histogram at the end of the collect period.

Note that the shortest collection interval GCM allows is 10 seconds, so setting both CollectPeriod and the batch timeout to 10 seconds is the lowest you should do.

**Reply by tyrone-anz** [reply]
Thanks @dashpole for looking into this.

Given we are using the `exact distribution` as aggregation, it is possible for us to get [Hi, Hi, Hi, Hi] (4 metric values with same labels) on one push. With and without batching, the issue for `duplicate timeseries is encountered` with those data. 

On trying out the 5-sec wait between the recording of same metric value, I got a different error.

Here is the summary of my testing. Hope it helps

```
// Recording of metric with different labels (with  batching on collector)
// <Intermittent error - happened only once, could be non-issue>
2021-10-11T16:44:27.960+1100    error   exporterhelper/queued_retry.go:254      Exporting failed. Try enabling retry_on_failure config option.  {"kind": "exporter", "name": "googlecloud", "error": "rpc error: code = Internal desc = One or more TimeSeries could not be written: Internal error encountered. Please retry after a few seconds. If internal errors persist, contact support at https://cloud.google.com/support/docs.: timeSeries[0-4]"}

// With recording of metric with same labels (with batching on collector)
2021-10-11T16:45:24.285+1100    error   exporterhelper/queued_retry.go:254      Exporting failed. Try enabling retry_on_failure config option.  {"kind": "exporter", "name": "googlecloud", "error": "rpc error: code = InvalidArgument desc = One or more TimeSeries could not be written: Field timeSeries[2] had an invalid value: Duplicate TimeSeries encountered. Only one point can be written per TimeSeries per request.: timeSeries[2]; Field timeSeries[3] had an invalid value: Duplicate TimeSeries encountered. Only one point can be written per TimeSeries per request.: timeSeries[3]; Field timeSeries[4] had an invalid value: Duplicate TimeSeries encountered. Only one point can be written per TimeSeries per request.: timeSeries[4]"}

// With recording of metric with different labels (no batching on collector)
// <No error>

// With recording of metric with same labels (no batching on collector)
2021-10-11T16:46:39.855+1100    error   exporterhelper/queued_retry.go:254      Exporting failed. Try enabling retry_on_failure config option.  {"kind": "exporter", "name": "googlecloud", "error": "rpc error: code = InvalidArgument desc = One or more TimeSeries could not be written: Field timeSeries[1] had an invalid value: Duplicate TimeSeries encountered. Only one point can be written per TimeSeries per request.: timeSeries[1]; Field timeSeries[2] had an invalid value: Duplicate TimeSeries encountered. Only one point can be written per TimeSeries per request.: timeSeries[2]; Field timeSeries[3] had an invalid value: Duplicate TimeSeries encountered. Only one point can be written per TimeSeries per request.: timeSeries[3]"}

// With a 5-sec wait between recording of metric with same labels (no batching on collector)
// First run seems fine, second run encounters the `max sampling error`
2021-10-11T16:59:57.939+1100    error   exporterhelper/queued_retry.go:254      Exporting failed. Try enabling retry_on_failure config option.  {"kind": "exporter", "name": "googlecloud", "error": "rpc error: code = InvalidArgument desc = One or more TimeSeries could not be written: One or more points were written more frequently than the maximum sampling period configured for the metric.: timeSeries[0]"}

```

**Reply by dashpole** [reply]
Unfortunately, the short answer is that you can't use an Exact distribution with google cloud right now.

**Reply by tyrone-anz** [reply]
Hi @dashpole, sorry me again. Would you know if the Exact distribution will be supported eventually? And if so, when do we hope to be able to use it?

**Reply by dashpole** [reply]
@tyrone-anz We don't have plans to support it right now.  Based on recent discussions in the opentelemetry-go SIG, the Exact distribution might be removed prior to the metrics API/SDK becoming stable, so we may not need to support it.  If it isn't removed, we would likely have limited support for the Exact distribution, since we can't send the same metric to GCM more than [once every 10 seconds](https://cloud.google.com/monitoring/quotas#custom_metrics_quotas).  This would mean we are forced to drop some of the recorded metrics in an Exact distribution if metrics are recorded more frequently than every 10 seconds.  At that point, I believe the Exact histogram becomes equivalent to a gauge with an CollectPeriod of 10 seconds.

**Reply by tyrone-anz** [reply]
Thanks for the prompt reply @dashpole.

**Reply by JCMais** [reply]
I am having a similar issue here, related to `Duplicate TimeSeries encountered`, is there any config on the otel collector that can help alleviate this?

**Reply by dashpole** [reply]
@JCMais there are a variety of issues that can cause Duplicate TimeSeries encountered.  The error usually means metrics from two instances of an application are being written to the same [monitored resource](https://cloud.google.com/monitoring/api/resources), and is solved by adding resource attributes to differentiate the two instances.  Feel free to open a separate discussion/issue, and we can figure out what the issue is.
